### ISSUE TITLE
Support Video Recovery on Sample Masters

### ISSUE THREAD

**Issue by suggestedfixes** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#269]
Currently, both sample masters have single thread of video stream providing to peers lazily created at receiving offer. Once it's there, the video thread acts as a singleton. One potential shortcoming is that, when this thread dies due to error, there isn't a mechanism to recover this video thread on subsequent offers.

One possible solution is to set mediaThreadStarted on FALSE on Cleanup of sendVideoPackets or sendGstreamerAudioVideo.
```
ATOMIC_STORE_BOOL(pSampleConfiguration->mediaThreadStarted, FALSE)
```
Any comments?

**Comment by suggestedfixes** [comment]
One shortcoming of the proposed approach is that the video sender might have died, but the audio has not. So we might end up creating duplicate audio threads on new offer.

**Comment by suggestedfixes** [comment]
To solve my last comment.
Can we split pSampleConfiguration->mediaThreadStarted, into videoThreadStarted and audioThreadStarted?

**Comment by MushMal** [comment]
@suggestedfixes perhaps there is some disconnect here. The peer connection is created on the main thread only for the viewer application. For the master, it's created on the receive message which is a separate thread. In either case, this has no bearing on the media pipeline. The master or viewer samples "simulate" the media pipeline by running a thread that feeds the samples.

The master sample should be able to handle new connection even if for some reason the previous session sender thread dies. There is a cleanup logic that handles expired sessions in the sample.

This said, the samples are just that - samples. Your commercial grade application might choose to handle things differently.

Could we try to get some specifics perhaps?

**Comment by suggestedfixes** [comment]
In Common.c, audio/video thread is started by this code block, this is creates a single video sending thread shared by all peers.
```

    if (!ATOMIC_LOAD_BOOL(&pSampleConfiguration->mediaThreadStarted)) {
        ATOMIC_STORE_BOOL(&pSampleConfiguration->mediaThreadStarted, TRUE);
        if (pSampleConfiguration->videoSource != NULL) {
            THREAD_CREATE(&pSampleConfiguration->videoSenderTid, pSampleConfiguration->videoSource,
                          (PVOID) pSampleConfiguration);
        }

        if (pSampleConfiguration->audioSource != NULL) {
            THREAD_CREATE(&pSampleConfiguration->audioSenderTid, pSampleConfiguration->audioSource,
                          (PVOID) pSampleConfiguration);
        }
    }

```

In gstsample, pSampleConfiguration->videoSource corresponds to sendGstreamerAudioVideo function. https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/samples/kvsWebRTCClientMasterGstreamerSample.c#L121

In normal conditions, this whole thread gets blocked by msg = gst_bus_timed_pop_filtered(bus, GST_CLOCK_TIME_NONE, GST_MESSAGE_ERROR | GST_MESSAGE_EOS);

In rare cases, when gstreamer dies, or there is a end of stream message by gstreamer, or there is an error. This only video/audio thread dies.

I'm proposing to set ATOMIC_STORE_BOOL(pSampleConfiguration->mediaThreadStarted, FALSE)
or splitting up mediaThreadStarted, so that on the next offer, we can recover this video/audio thread.

**Comment by MushMal** [comment]
This code is executed only for the master to start off the handler threads. You are right, in case of gstreamer it will block until gstreamer exits gracefully. The mediaThreadStarted guards only the start sequence. It won't work neatly to re-start the terminated handler thread. There needs to be a better way to handle each of the handler threads in case they die. The original sample for gstreamer left this out of scope as we anticipate the gstreamer wouldn't simply crash. Perhaps we do need to handle the restarting or perhaps a graceful shutdown of the session on the master in case of this type of errors so the higher level logic can re-create. We really need to think through.

There is an issue tracking this https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/issues/205

No ETA though on when we can fix.

**Comment by MushMal** [comment]
I am going to close this issue in favor of #205
